recipes/eleuther_eval.py

```diff
@@ -0,0 +1,189 @@
+        # Note on add_bos flag: setting to False as this gives better results, for example
+        # +1% on truthfulqa_mc2 with a LoRA finetune. lit-gpt also sets this to False,
+        # see https://github.com/Lightning-AI/lit-gpt/blob/main/eval/lm_eval_harness.py#L66,
+        # though notably fast-gpt does the opposite
+        # https://github.com/pytorch-labs/gpt-fast/blob/main/eval.py#L123.
+        return self._tokenizer.encode(text=string, add_bos=False, add_eos=False)
+
+    def tok_decode(self, tokens, **kwargs):
+        return self._tokenizer.decode(tokens)
+
+    def _model_call(self, inps):
+        return self._model(inps)
+
+    def _model_generate(self, *args, **kwargs):
+        raise RuntimeError(
+            "This recipe does not currently support tasks that evaluate free generation,"
+            "e.g. `truthfulqa_gen` or `bigbench_color_generate_until`."
+        )
+
+
+_DEFAULT_TASKS = ["hellaswag"]
+
+
+class EleutherEvalRecipe(EvalRecipeInterface):
+    """This recipe runs evaluation on a trained model using EleutherAI's eval harness."""
+
+    def __init__(self, cfg: DictConfig) -> None:
+        self._cfg = cfg
+
+    def load_checkpoint(self, cfg: DictConfig) -> Dict[str, Any]:
+        checkpointer = config.instantiate(cfg)
+        checkpoint_dict = checkpointer.load_checkpoint()
+        return checkpoint_dict
+
+    def setup(self) -> None:
+        self._device = utils.get_device(device=self._cfg.device)
+        self._dtype = utils.get_dtype(dtype=self._cfg.dtype)
+        self._limit = self._cfg.limit
+        self._tasks = list(self._cfg.tasks or _DEFAULT_TASKS)
+
+        seed = utils.set_seed(seed=self._cfg.seed)
+        logger.info(f"Random seed set to {seed}.")
+
+        ckpt_dict = self.load_checkpoint(self._cfg.checkpointer)
+        self._model = self._setup_model(
+            cfg_model=self._cfg.model,
+            model_state_dict=ckpt_dict[utils.MODEL_KEY],
+        )
+        self._tokenizer = config.instantiate(self._cfg.tokenizer)
+        logger.info("Tokenizer is initialized from file.")
+
+    def _setup_model(
+        self,
+        cfg_model: DictConfig,
+        model_state_dict: Dict[str, Any],
+    ) -> nn.Module:
+        with utils.set_default_dtype(self._dtype), self._device:
+            model = config.instantiate(cfg_model)
+
+        model.load_state_dict(model_state_dict)
+
+        # Validate model was loaded in with the expected dtype.
+        utils.validate_expected_param_dtype(model, dtype=self._dtype)
+        logger.info(f"Model is initialized with precision {self._dtype}.")
+        return model
+
+    @torch.no_grad()
+    def evaluate(self) -> None:
+        t1 = time.time()
+
+        model_eval_wrapper = _EvalWrapper(
+            self._model,
+            self._tokenizer,
+            device=self._device,
+            max_seq_length=self._cfg.max_seq_length,
+        )
+
+        # Task initialization API changed between v0.4.1 and 0.4.2
```

Comment:
Copied this from gpt-fast

Reply:
Can we give a bit more detail here? And maybe an explicit type of exception?